print-geonet.c

```diff
@@ -26,232 +27,481 @@
 
-
 /*
-   ETSI TS 102 636-5-1 V1.1.1 (2011-02)
-   Intelligent Transport Systems (ITS); Vehicular Communications; GeoNetworking;
-   Part 5: Transport Protocols; Sub-part 1: Basic Transport Protocol
+   ETSI EN 302 636-5-1 V2.2.1 (2019-05)
+   Intelligent Transport Systems (ITS);
+   Vehicular Communications;
+   GeoNetworking;
+   Part 5: Transport Protocols;
+   Sub-part 1: Basic Transport Protocol
 
-   ETSI TS 102 636-4-1 V1.1.1 (2011-06)
-   Intelligent Transport Systems (ITS); Vehicular communications; GeoNetworking;
+   ETSI EN 302 636-4-1 V1.4.1 (2020-01)
+   Intelligent Transport Systems (ITS);
+   Vehicular Communications;
+   GeoNetworking;
    Part 4: Geographical addressing and forwarding for point-to-point and point-to-multipoint communications;
-   Sub-part 1: Media-Independent Functionality
+   Sub-part 1: Media-Independent Functionality;
+   Release 2
 */
 
-#define GEONET_ADDR_LEN 8
-
-static const struct tok msg_type_values[] = {
-	{   0, "CAM" },
-	{   1, "DENM" },
-	{ 101, "TPEGM" },
-	{ 102, "TSPDM" },
-	{ 103, "VPM" },
-	{ 104, "SRM" },
-	{ 105, "SLAM" },
-	{ 106, "ecoCAM" },
-	{ 107, "ITM" },
-	{ 150, "SA" },
-	{   0, NULL }
+/*Specific Definitions*/
+#define NDO_V_FLAG_FIRST_DEBUG_LEVEL 1
+#define NDO_V_FLAG_SECOND_DEBUG_LEVEL 2
+#define NDO_V_FLAG_THIRD_DEBUG_LEVEL 3
+
+/*Bit-Wise Definitions*/
+#define ONE_BYTE 8
+#define TWO_BYTES 16
+#define THREE_BYTES 24
+#define FOUR_BYTES 32
+#define FIVE_BYTES 40
+#define SIX_BYTES 48
+#define SEVEN_BYTES 56
+#define EIGHT_BYTES 64
+
+#define ONE_BIT_MASK 0x01
+#define TWO_BITS_MASK 0x03
+#define THREE_BITS_MASK 0x07
+#define FOUR_BITS_MASK 0x0F
+#define FIVE_BITS_MASK 0x1F
+#define SIX_BITS_MASK 0x3F
+#define SEVEN_BITS_MASK 0x7F
+#define EIGHT_BITS_MASK 0xFF
+#define TEN_BITS_MASK 0x3FF
+#define SIXTEEN_BITS_MASK 0xFFFF
+#define FORTY_EIGHT_BITS_MASK 0xFFFFFFFFFFFF
+
+/* GeoNetworking Definitons*/
+
+/* GeoNetworking Basic Header Definitions*/
+#define GN_BASIC_HEADER_MINIMUM_PACKET_LENGTH 4
+
+#define IMPLEMENTED_GN_VERSIONS_NUM 1
+static const u_int implemented_gn_versions[IMPLEMENTED_GN_VERSIONS_NUM] = {1};
+
+#define NH_COMMONHEADER 1
+#define NH_SECUREDPACKET 2
+#define IMPLEMENTED_GN_NEXT_HEADERS_NUM 1
+static const u_int implemented_gn_nh_headers[IMPLEMENTED_GN_NEXT_HEADERS_NUM] = {
+	NH_COMMONHEADER};
+static const struct tok basic_header_next_header_values[] = {
+	{0, "Any"},
+	{NH_COMMONHEADER, "CommonHeader"},
+	{NH_SECUREDPACKET, "SecuredPacket"},
 };
 
-static void
-print_btp_body(netdissect_options *ndo,
-	       const u_char *bp)
-{
-	u_int msg_type;
+#define HT_BEACON 1
+#define HT_TSB 5
+#define HT_TSB_SHB 0
+#define HT_TSB_MULTI_HOP 1
+#define ELAPSED_SECONDS 5
+
+#define IMPLEMENTED_GN_HEADER_TYPES_NUM 2
+static const u_int implemented_gn_header_types[IMPLEMENTED_GN_HEADER_TYPES_NUM] = {
+	HT_BEACON,
+	HT_TSB};
+
+static const struct tok common_header_next_header_values[] = {
+	{0, "Any"},
+	{1, "BTP-A"},
+	{2, "BTP-B"},
+	{3, "IPv6"},
```

Comment:
Please either terminate every array of `struct tok` with a `{ 0, NULL }` record or switch to `struct uint_tokary` and `uint2tokary()`, otherwise this code will crash.